Assets/Mirror/Runtime/ProfilerData.cs

```diff
@@ -0,0 +1,109 @@
+using System;
+using System.Diagnostics;
+using UnityEngine.Events;
+
+namespace Mirror
+{
+    /// <summary>
+    /// Provides profiling information from mirror
+    /// A profiler can subscribe to these events and
+    /// present the data in a friendly way to the user
+    /// </summary>
+    public static class ProfilerData
+    {
+        #region Out messages
+
+        /// <summary>
+        /// Describes an outgoing message
+        /// </summary>
+        public readonly struct OutMessage
+        {
+            /// <summary>
+            /// The message being sent
+            /// </summary>
+            public readonly IMessageBase message;
+            /// <summary>
+            /// channel through which the message was sent
+            /// </summary>
+            public readonly int channel;
+            /// <summary>
+            /// how big was the message (does not include transport headers)
+            /// </summary>
+            public readonly int bytes;
+            /// <summary>
+            /// How many connections was the message sent to
+            /// If an object has a lot of observers this count could be high
+            /// </summary>
+            public readonly int count;
+
+            internal OutMessage(IMessageBase message, int channel, int bytes, int count)
+            {
+                this.message = message;
+                this.channel = channel;
+                this.bytes = bytes;
+                this.count = count;
+            }
+        }
+
+        public class OutMessageUnityEvent : UnityEvent<OutMessage> { };
```

Comment:
Are UnityEvents just as fast as native C# events?  This needs to be ultra fast.

Reply:
UnityEvents were invented because unlike native C# events they do no allocations.

See here for in depth analysis
https://jacksondunstan.com/articles/3335